fix: profile 팔로잉 팔로워 모달 문제 해결

Yoon-Hae-Min opened the pull request:
## 관련 이슈
<!-- close #이슈번호 -->
close #155 

## 예상 리뷰 시간
10-min

## 추가 또는 변경사항
<!-- 구체적으로 작성 부탁드립니다. -->
1.  타 유저 팔로우 버튼 클릭 시 팔로워 카운트가 오르지만 팔로워 모달에는 업데이트가 안되어있는 문제 해결
2.  마이 페이지에서 팔로워 또는 팔로잉 유저를 삭제했을 때 모달에서는 사라지지만 카운트가 내려가지 않는 문제 해결
3.  마이 페이지 삭제 버튼을 팔로우 언팔로우 버튼으로 변경
4. 언팔로우 팔로우 버튼을 클릭해도 modal창에서 사라지지 않음 (페이지를 새로 고침해야 반영이됨)
5. unfollow api 이름 delete로 변경
6. 팔로우 팔로워시에 로딩 추가

8/10추가
~~1. modal에서 팔로우 언팔로우 버튼을 누를시 다른쪽 modal 데이터를 다시 불러오게 하였습니다 사용자는 로딩창 없이 팔로우 팔로워 모달을 둘다 볼수 있습니다.~~
2. 나 자신을 팔로우하지 못하게 isPermission옵션을 userId와 비교하였습니다.

## 스크린샷
![1](https://user-images.githubusercontent.com/49224104/183592419-8462dcc0-b031-417d-affc-b9c778467457.gif)
팔로우하고 모달을 껐다 켜도 남아있는 모습
![2](https://user-images.githubusercontent.com/49224104/183592447-b9bc1d5f-cdaf-44e2-b345-2c81ee27df71.gif)
타 페이지에서 팔로우 버튼을 클릭할시에 유저정보를 새로 받아옴


## 기타
<!-- 작업 중 있언던 것들을 자유롭게 작성합니다. -->
follow mustation 구문을 각 컴포넌트로 이동하였습니다 followList같은 경우에는 카드 하나당 하나의 loading과 state가 필요하여mutation을 새로 만들었고 그러다보니 page단에서 mutation을 가지고 있을 필요가없어 이동하였습니다.



<br/>

<br/>

<br/>

<br/>

<br/>

8/11
대대적으로 계속 오류가 발생해서 동작 테스트 진행을 해보려 합니다.

1. 팔로워 모달 클릭후 팔로우 O
2. 팔로워 모달 클릭후 언팔로우 O
3. 팔로워 모달 클릭후 팔로우 클릭후 팔로잉 모달 반영 확인
4. 팔로워 모달 클릭후 언팔로우 클릭후 팔로잉 모달 반영 확인 O
5. 팔로워 모달 클릭후 팔로우 클릭후 팔로잉 모달에서 언팔로우 후 팔로워 모달에서 반영 확인 O
6. 팔로워 모달 클릭후 언팔로우 클릭후 팔로잉 모달에서 팔로우 후 팔로워 모달에서 반영 확인 
7. 팔로잉 모달 클릭후 팔로우 O
8. 팔로잉 모달 클릭후 언팔로우 O
9. 팔로잉 모달 클릭후 팔로우 클릭후 팔로워 모달 반영 확인(타 유저 페이지에서) O
10. 팔로잉 모달 클릭후 언팔로우 클릭후 팔로워 모달 반영 확인(타유저 페이지에서) O
11. 팔로잉 모달 클릭후 팔로우 클릭후 팔로워 모달에서 언팔로우 후 팔로잉 모달에서 반영 확인 O
12. 팔로잉 모달 클릭후 언팔로우 클릭후 팔로워 모달에서 팔로우 후 팔로잉 모달에서 반영 확인 O

<br/>

3,6번의 내용 중 현재 내 페이지에서 어떤 유저가 나를 팔로우 하지만 내가 그 유저를 팔로우 하지 않을때 불가능합니다. (유저가 팔로워에만 있을때 팔로잉에 추가가 안됨)
팔로워에서 팔로우를 누르면 팔로잉 카운트는 올라가지만 팔로잉 modal 목록에는 card로 반영되지 않습니다.
하지만 인피니티 스크롤링시에 데이터를 가져오기 때문에 마지막 데이터로 반영될 것이라고 생각합니다.

<br/>

ps) 다만 전부 가져와 버리면 인피니티 스크롤링이 끝나 데이터를 갱신할수 없음 (논의필요)
현재 생각으로는 반영이 안돼도 괜찮지 않을까 생각합니다 팔로워 목록에서 card가 남아있음+ 팔로잉 숫자로는 카운트됨 + 새로고침시에 적용이됨+ 인피니티 스크롤링이 끝나는 경우만 발생 이 4개를 고려해봤을때는 크게 ux를 해치는건 없을거라고 생각합니다.

Yoon-Hae-Min committed:
refactor: api prefix 변경

Yoon-Hae-Min committed:
feat: modal 팔로우 팔로워 버튼으로 교체

Yoon-Hae-Min committed:
feat: 팔로우 팔로워 modal 버튼 로직 변경

Yoon-Hae-Min committed:
feat: 프로필페이지 팔로우 언팔로우 버튼 로직 수정

Yoon-Hae-Min committed:
feat: 타 유저 팔로우 버튼 클릭 시 팔로워 모달에도 업데이트

Yoon-Hae-Min committed:
feat: 팔로우 팔로워 modal에서도 팔로우 언팔로우 할수 있게 추가

Yoon-Hae-Min committed:
Merge branch 'develop' into feat/155-fix-following-follower-modal

# Conflicts:
#	breaking-front/src/mocks/profileHandlers.js

kangju2000 commented on breaking-front/src/components/FollowCard/FollowCard.js:
```diff
@@ -1,15 +1,24 @@
+  const queryClient = useQueryClient();
+  const [isFollow, setIsFollow] = useState(profileData.isFollowing);
+  const { mutate: UnFollow, isLoading: isUnFollowLoading } =
+    useMutation(deleteUnFollow);
```
이런 식으로 가면 61번 줄에서 mutate하는 코드가 줄어 가독성이 좋을 것 같습니다. ProfileFollowButton의 useMutation 부분도 이렇게 수정하면 좋겠습니다!
Follow,
```suggestion
  const { mutate: UnFollow, isLoading: isUnFollowLoading } =
    useMutation(deleteUnFollow, {
      onSuccess: () => {
                    queryClient.invalidateQueries('profile');
                    toggleIsFollow();
                  },
      onError: () => {
                    //에러처리
                  },
    });
```

kangju2000 commented on breaking-front/src/pages/Profile/Profile.js:
```diff
@@ -18,14 +17,12 @@ import useProfileWrittenPost from 'hooks/queries/useProfileWrittenPost';
 import { useNavigate, useParams } from 'react-router-dom';
 import ImageUrlConverter from 'utils/ImageUrlConverter';
 import ProfileFollowButton from './units/ProfileFollowButton';
 import ProfileTabPanel from './units/ProfileTabPanel';
```
절대 경로 수정해주세용

kangju2000 commented on breaking-front/src/pages/Profile/units/ProfileFollowButton.styles.js:
```diff
@@ -1,8 +1,15 @@
+import ReactLoading from 'react-loading';
 
 export const FollowButton = styled(Button)`
   margin-left: 20px;
-  padding: 10px 15px;
+  padding: 5px;
+  width: 80px;
+  height: 40px;
   font-size: 18px;
 `;
```
코딩 컨벤션 맞춰주세요.

kangju2000 reviewed (changes requested):
본인 프로필 페이지에서 팔로워에 있는 유저를 팔로우 했을 때 팔로잉에 유저 카운트가 1 증가하지만, 팔로잉 modal 안에 그 유저가 포함되어있지 않습니다.
modal을 닫고 새로 열었을 때 list를 업데이트 해야합니다.

Yoon-Hae-Min committed:
feat: modal 파일 분리 및 팔로우 팔로잉 버튼 누를시에 반대쪽 modal창 리랜더링

Yoon-Hae-Min committed:
feat: 나 자신을 팔로우 하지 못하게 설정

Yoon-Hae-Min committed:
style: 절대경로로 수정

Yoon-Hae-Min committed:
feat: modal창에서 팔로우 언팔로우 로직 변경

1. 모달창을 2개로 분리
2. 각 데이터를 state로 관리하고 팔로우 언팔로우시 state를 변경하여 반대쪽 modal에도 적용값 반영

Yoon-Hae-Min committed:
refactor: 팔로워 팔로잉 모달을 공통적인 컴포넌트로 묶음

Yoon-Hae-Min committed:
feat: 숫자를 보여주어야 하는 부분에 numberFormatter를 적용

kangju2000 commented on breaking-front/src/components/FollowCard/FollowCard.js:
```diff
@@ -11,18 +12,30 @@ export default function FollowCard({
   const theme = useTheme();
   const queryClient = useQueryClient();
-  const [isFollow, setIsFollow] = useState(profileData.isFollowing);
+  const ChangeIsFollowFelid = (pre) => {
```
Felid는 오타인 것 같은데 수정 바랍니다.

kangju2000 commented on breaking-front/src/pages/Profile/units/ProfileFollowModal.js:
```diff
@@ -1,66 +1,105 @@
-  toggleModal,
 }) => {
   const navigate = useNavigate();
   const myData = useContext(UserInformationContext);
```
저는 useContext에서 가져온 유저 정보를 `userData` 라고 작명하여 사용했는데 둘 중 통일해야할 것 같습니다.

kangju2000 commented on breaking-front/src/pages/Profile/Profile.js:
```diff
@@ -63,83 +50,25 @@ const Profile = () => {
+        toggleFollowerModal={toggleFollowerModal}
+        toggleFollowingModal={toggleFollowingModal}
+        userId={userId}
+      />
```
ProfileFollowmodals가 아닌 ProfileFollowModal로 써야합니다.
그리고 ProfileFollowModal을 컴포넌트로 분리하는 것 보다 Profile.js에 직접 풀어쓰는게 더 좋아보입니다.

Yoon-Hae-Min committed:
style: 오타 및 코딩 컨벤션 추가

Yoon-Hae-Min committed:
style: 파일명 단수로변경 및 변수명 수정

- modal은 동사에서 유래된거라 복수로 쓸수 없음
- 공통인 modal을 common으로 묶음
- context로 가져오는 로그인정보를 userData로 정정

Yoon-Hae-Min committed:
style: List를 followList로 변수명 변경

kangju2000 reviewed (approved):
고생하셨습니다!
